src/bindings/g_object/object.cr

```diff
@@ -510,8 +510,23 @@ module GObject
         def emit({{ signature.args.splat }})
           {% for arg in signature.args %}
             {% resolved_type = arg.type.resolve %}
-            {% if resolved_type < Reference && resolved_type != ::String && !(resolved_type < ::GObject::Object) %}
-              {{ arg.var }} = {{ arg.var }}.as(Void*)
+            # If the parameter is a pure Crystal object, i.e. it doesn't inherits from any
+            # object generated by GI-Crystal, AND it's a reference, we just pass a pointer
+            # to the crystal object to the signal, otherwise the `to_unsafe`.
+            #
+            # Note that if the object is a pure crystal object and has the #to_unsafe method
+            # we can't use it, since we can't reconstruct this object with that pointer returned by
+            # the #to_unsafe call.
+            {% if resolved_type != ::String && resolved_type < Reference %}
+              {% pure_crystal = true %}
+              {% for ancestor in resolved_type.ancestors %}
+                {% if ancestor.annotation(GICrystal::GeneratedWrapper) %}
+                  {% pure_crystal = false %}
+                {% end %}
+              {% end %}
```

Comment:
runtime efficiency isn't an issue since this probably just add few nano seconds at compile time, but I agree this ugly loop is a poor design made by someone that just wanted to have this done :sweat_smile: 

A module would be much more elegant.